### ISSUE TITLE
( lazyvec1 - lazyvec1.t() ) @ lazyvec2 fails with: 'NoneType' object has no attribute 'replace'

### ISSUE THREAD

**Issue by joanglaunes** [getkeops/keops#211]
### Discussed in https://github.com/getkeops/keops/discussions/210

<div type='discussions-op-text'>

<sup>Originally posted by **tvercaut** February  3, 2022</sup>
Hi,

Thanks for the nice work. I am not sure if I am using the library correctly as it is the first time I try it  but the following behaviour looks counter-intuitive to me. Can someone tell me if it's a bug or a misunderstanding from my part?

For two vectors `u` and `v` of size (Nx1) I would like to compute `(u_l - u_l.T) @ v` in a lazy fashion. Doing this directly with `u_l` being the lazy vector corresponding to the concrete vector `u`, and `v` being a concrete vector fails with the following error:
```
'NoneType' object has no attribute 'replace'
```

The workaround I found was to create a second lazy vector from the transposed concrete vector `u.t()` however this also was counter-intuitive since `LazyTensor(u.t(),axis=1)` is reported to be of size 1x1xN rather than 1xN. The second workaround I found was to create the lazy vector from a 3D tensor of size 1xNx1.

Let me know what I might be missing.

Here is an example (also on [colab](https://colab.research.google.com/drive/15WAK3HtSQuriW2ohFuMSGFOEFTvPdEBr?usp=sharing)):
```python
import torch
print(torch.__version__)

torchdevice = torch.device('cpu')
if torch.cuda.is_available():
  torchdevice = torch.device('cuda')
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))
print('Running on ' + str(torchdevice))

!pip install pykeops[colab]
import pykeops
print('pykeops version:',pykeops.__version__)

from pykeops.torch import LazyTensor

# Generate 2 vector u and v
u = torch.normal(0, 1, size=(5,1), device=torchdevice)
v = torch.normal(0, 1, size=(5,), device=torchdevice)

# Compute (u - u.T) @ v in a dense fashion
P = u - u.t()
Pv = P @ v
print('Pv: ',Pv)

# Compute (u - u.T) @ v in a lazy fashion
u_l = LazyTensor(u,axis=0)
print('u_l: ', u_l)

# Using the transpose of the lazy tensor doesn't work here
#print('u_l.t(): ', u_l.t())
#P_l =  u_l - u_l.t()
#print('P_l:', P_l)

# We thus go for transposing the concrete vector and creating a new lazy tensor
# However, it seems to fails when provided as a 2D matrix of size 1 x N
#ut_l = LazyTensor(u.t(),axis=1)
print('LazyTensor(u.t(),axis=1): ',LazyTensor(u.t(),axis=1))
ut_l = LazyTensor(u.t()[...,None])
print('ut_l: ', ut_l)

P_l =  u_l - ut_l
print('P_l:', P_l)

Pv_froml = P_l @ v
print('Pv_froml: ',Pv_froml)

# Putting this at the end just to highlight that this fails
( u_l - u_l.t() ) @ v
```</div>

**Comment by joanglaunes** [comment]
Hello Tom,

Thanks for your interest in our library.
The first error is a clear bug from the transpose method `.t()` in the LazyTensor class. I have now fixed it in the master branch. So in colab for example your script works now if you install the master branch version with 
```
!pip install cmake==3.18
!pip install git+https://github.com/getkeops/keops.git@master
```
Thank you for noticing and mentioning this bug.

Now for the second problem you mention, it is not really a bug but rather an odd behavior due to the way we interpret dimensions in KeOps : when you convert a 2D tensor to LazyTensor the input shape is interpreted as (N,D) (sample size and dimension of each point/vector), so with the option axis=0 it will give a LazyTensor of shape (N,1,D) (the N dimension is put in first position), and with axis=1 it will give shape (1,N,D). In your case you input u.t() which is of shape (1,5), so it will interpret as N=1 and D=5 and with axis=0 it gives a (1,1,5) LazyTensor. I know this is really confusing in the specific use case you have... We already thought about changing this behaviour or adding another class similar to LazyTensor but with clearer interface, but it not done yet.

**Comment by tvercaut** [comment]
Many thanks @joanglaunes. That makes sense now.

**Comment by joanglaunes** [comment]
Hello @jeanfeydy ,
Actually for the bug with the first example (test from docstring), it was a small typo in the code. I have done the fix in master now. Can you try your optimal transport scripts again and see if it solves the issue also for them ?

**Comment by jeanfeydy** [comment]
Hi @joanglaunes,
Yes, as far as I can tell, things run smoothly now: thanks a lot!
I am still fixing/documenting bugs as I encounter them. 
I will let you know when everything is OK for the release :-)
